### ISSUE TITLE
The C++ vit h.onnx failed to convert trt

### ISSUE THREAD

**Issue by guanhaiye** [mingj2021/segment-anything-tensorrt#13]
Because vit_h is greater than 2G, there are many files to export onnx, how can I export trt in C++? Has the host tried to export VIT_H. engine in C++

**Comment by mingj2021** [comment]
我的思路，将 encoder  和 decoder  按照block 拆成合适的submodel 先转onnx 在转trt  ，同时把 FP32 变成 FP16 ，大模型要拆分submodel 去部署，这是我的理解，稍后我会去试试的

**Comment by YzbDoMore** [comment]
> 我的思路，将 encoder 和 decoder 按照block 拆成合适的submodel 先转onnx 在转trt ，同时把 FP32 变成 FP16 ，大模型要拆分submodel 去部署，这是我的理解，稍后我会去试试的
如果成功，能分享一下代码吗

**Comment by guanhaiye** [comment]
> 我的思路，将 encoder 和 decoder 按照block 拆成合适的submodel 先转onnx 在转trt ，同时把 FP32 变成 FP16 ，大模型要拆分submodel 去部署，这是我的理解，稍后我会去试试的

If successful, can you share the complete code

**Comment by mingj2021** [comment]
🆗

**Comment by mingj2021** [comment]
vim_h model 已经转换成功，将embedings model 一分为二，进行推理，在python环境验证成功，c++端 转trt时，需要进一步验证，可以不设置FP16 值，代码一分享。

**Comment by mingj2021** [comment]
在 GPU 3080 Ti 测试正常

**Comment by guanhaiye** [comment]
> 在 GPU 3080 Ti 测试正常

tks

**Comment by guanhaiye** [comment]
> vim_h model 已经转换成功，将embedings model 一分为二，进行推理，在python环境验证成功，c++端 转trt时，需要进一步验证，可以不设置FP16 值，代码一分享。

The embedings model will be divided into two, reasoning, and successfully verified in the python environment. Can you share the code of the main python side

**Comment by LightSun** [comment]
同求

**Comment by guanhaiye** [comment]
> vim_h model 已经转换成功，将embedings model 一分为二，进行推理，在python环境验证成功，c++端 转trt时，需要进一步验证，可以不设置FP16 值，代码一分享。

The embedings model will be divided into two, reasoning, and successfully verified in the python environment. Can you share the code of the main python side

**Comment by mingj2021** [comment]
class embedding_model_part_1(nn.Module):
    def __init__(self) :
        super().__init__()
        sam_checkpoint = "weights/sam_vit_h_4b8939.pth"
        model_type = "vit_h"
        device = "cpu"
        self.sam = sam_model_registry[model_type](checkpoint=sam_checkpoint)
        self.sam.to(device=device)

    def forward(self, x):
        x = self.sam.image_encoder.patch_embed(x)
        if self.sam.image_encoder.pos_embed is not None:
            x = x + self.sam.image_encoder.pos_embed

        val_1 = len(self.sam.image_encoder.blocks) // 2
        for blk in self.sam.image_encoder.blocks[0:val_1]:
            x = blk(x)

        return x
    
class embedding_model_part_2(nn.Module):
    def __init__(self) :
        super().__init__()
        sam_checkpoint = "weights/sam_vit_h_4b8939.pth"
        model_type = "vit_h"
        device = "cpu"
        self.sam = sam_model_registry[model_type](checkpoint=sam_checkpoint)
        self.sam.to(device=device)

    def forward(self, x):
        val_1 = len(self.sam.image_encoder.blocks) // 2
        for blk in self.sam.image_encoder.blocks[val_1:]:
            x = blk(x)
        x = self.sam.image_encoder.neck(x.permute(0, 3, 1, 2))
        return x

def export_embedding_model_part_1():
    device = "cpu"
    model_type = "vit_h"
    model = embedding_model_part_1()
    # image_encoder = EmbeddingOnnxModel(sam)
    image = cv2.imread('notebooks/images/truck.jpg')
    target_length = model.sam.image_encoder.img_size
    pixel_mean = model.sam.pixel_mean 
    pixel_std = model.sam.pixel_std
    img_size = model.sam.image_encoder.img_size
    inputs = pre_processing(image, target_length, device,pixel_mean,pixel_std,img_size)
    onnx_model_path = model_type+"_"+"part_1_embedding.onnx"
    dummy_inputs = {
    "images": inputs
}
    output_names = ["image_embeddings_part_1"]
    with warnings.catch_warnings():
        warnings.filterwarnings("ignore", category=torch.jit.TracerWarning)
        warnings.filterwarnings("ignore", category=UserWarning)
        with open(onnx_model_path, "wb") as f:
            torch.onnx.export(
                model,
                tuple(dummy_inputs.values()),
                f,
                export_params=True,
                verbose=False,
                opset_version=13,
                do_constant_folding=True,
                input_names=list(dummy_inputs.keys()),
                output_names=output_names,
                # dynamic_axes=dynamic_axes,
            )   

def export_embedding_model_part_2():
    device = "cpu"
    model_type = "vit_h"
    model = embedding_model_part_2()
    
    inputs = torch.randn(1, 64, 64, 1280)
    onnx_model_path = model_type+"_"+"part_2_embedding.onnx"
    dummy_inputs = {
    "image_embeddings_part_1": inputs
}
    output_names = ["image_embeddings_part_2"]
    with warnings.catch_warnings():
        warnings.filterwarnings("ignore", category=torch.jit.TracerWarning)
        warnings.filterwarnings("ignore", category=UserWarning)
        with open(onnx_model_path, "wb") as f:
            torch.onnx.export(
                model,
                tuple(dummy_inputs.values()),
                f,
                export_params=True,
                verbose=False,
                opset_version=13,
                do_constant_folding=True,
                input_names=list(dummy_inputs.keys()),
                output_names=output_names,
                # dynamic_axes=dynamic_axes,
            )   
        
def export_sam_h_model():
    from segment_anything.utils.onnx import SamOnnxModel
    checkpoint = "weights/sam_vit_h_4b8939.pth"
    model_type = "vit_h"
    sam = sam_model_registry[model_type](checkpoint=checkpoint)
    onnx_model_path = "sam_h_decoder_onnx.onnx"

    onnx_model = SamOnnxModel(sam, return_single_mask=True)

    dynamic_axes = {
        "point_coords": {1: "num_points"},
        "point_labels": {1: "num_points"},
    }

    embed_dim = sam.prompt_encoder.embed_dim
    embed_size = sam.prompt_encoder.image_embedding_size
    mask_input_size = [4 * x for x in embed_size]
    dummy_inputs = {
        "image_embeddings": torch.randn(1, embed_dim, *embed_size, dtype=torch.float),
        "point_coords": torch.randint(low=0, high=1024, size=(1, 5, 2), dtype=torch.float),
        "point_labels": torch.randint(low=0, high=4, size=(1, 5), dtype=torch.float),
        "mask_input": torch.randn(1, 1, *mask_input_size, dtype=torch.float),
        "has_mask_input": torch.tensor([1], dtype=torch.float),
        # "orig_im_size": torch.tensor([1500, 2250], dtype=torch.float),
    }
    # output_names = ["masks", "iou_predictions", "low_res_masks"]
    output_names = ["masks", "scores"]

    with warnings.catch_warnings():
        warnings.filterwarnings("ignore", category=torch.jit.TracerWarning)
        warnings.filterwarnings("ignore", category=UserWarning)
        with open(onnx_model_path, "wb") as f:
            torch.onnx.export(
                onnx_model,
                tuple(dummy_inputs.values()),
                f,
                export_params=True,
                verbose=False,
                opset_version=13,
                do_constant_folding=True,
                input_names=list(dummy_inputs.keys()),
                output_names=output_names,
                dynamic_axes=dynamic_axes,
            )  

def onnx_model_example2():
    import os
    ort_session_embedding_part_1 = onnxruntime.InferenceSession('vit_h_part_1_embedding.onnx',providers=['CPUExecutionProvider'])
    ort_session_embedding_part_2 = onnxruntime.InferenceSession('vit_h_part_2_embedding.onnx',providers=['CPUExecutionProvider'])
    ort_session_sam = onnxruntime.InferenceSession('sam_h_decoder_onnx.onnx',providers=['CPUExecutionProvider'])

    image = cv2.imread('notebooks/images/truck.jpg')
    image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
    image2 = image.copy()
    prompt_embed_dim = 256
    image_size = 1024
    vit_patch_size = 16
    image_embedding_size = image_size // vit_patch_size
    target_length = image_size
    pixel_mean=[123.675, 116.28, 103.53],
    pixel_std=[58.395, 57.12, 57.375]
    pixel_mean = torch.Tensor(pixel_mean).view(-1, 1, 1)
    pixel_std = torch.Tensor(pixel_std).view(-1, 1, 1)
    device = "cpu"
    inputs = pre_processing(image, target_length, device,pixel_mean,pixel_std,image_size)
    ort_inputs = {
    "images": inputs.cpu().numpy()
    }
    image_embeddings = ort_session_embedding_part_1.run(None, ort_inputs)[0]
    ort_inputs = {
    "image_embeddings_part_1": image_embeddings
    }
    image_embeddings = ort_session_embedding_part_2.run(None, ort_inputs)[0]

    input_point = np.array([[784, 379]])
    input_label = np.array([1])

    onnx_coord = np.concatenate([input_point, np.array([[0.0, 0.0]])], axis=0)[None, :, :]
    onnx_label = np.concatenate([input_label, np.array([-1])], axis=0)[None, :].astype(np.float32)
    from segment_anything.utils.transforms import ResizeLongestSide
    transf = ResizeLongestSide(image_size)
    onnx_coord = transf.apply_coords(onnx_coord, image.shape[:2]).astype(np.float32)
    onnx_mask_input = np.zeros((1, 1, 256, 256), dtype=np.float32)
    onnx_has_mask_input = np.zeros(1, dtype=np.float32)

    ort_inputs = {
        "image_embeddings": image_embeddings,
        "point_coords": onnx_coord,
        "point_labels": onnx_label,
        "mask_input": onnx_mask_input,
        "has_mask_input": onnx_has_mask_input,
        # "orig_im_size": np.array(image.shape[:2], dtype=np.float32)
    }

    masks, _ = ort_session_sam.run(None, ort_inputs)

    from segment_anything.utils.onnx import SamOnnxModel
    checkpoint = "weights/sam_vit_h_4b8939.pth"
    model_type = "vit_h"
    sam = sam_model_registry[model_type](checkpoint=checkpoint)

    onnx_model = SamOnnxModel(sam, return_single_mask=True)
    masks = onnx_model.mask_postprocessing(torch.as_tensor(masks), torch.as_tensor(image.shape[:2]))
    masks = masks > 0.0
    plt.figure(figsize=(10, 10))
    plt.imshow(image)
    show_mask(masks, plt.gca())
    # show_box(input_box, plt.gca())
    show_points(input_point, input_label, plt.gca())
    plt.axis('off')
    plt.savefig('demo.png')
    plt.show()


if __name__ == '__main__':
    with torch.no_grad():
        # export_embedding_model()
        # export_sam_model()
        # onnx_model_example()
        # export_engine_image_encoder()
        # export_engine_prompt_encoder_and_mask_decoder()
        export_embedding_model_part_1()
        export_embedding_model_part_2()
        export_sam_h_model()
        onnx_model_example2()

**Comment by mingj2021** [comment]
请看一下 how_to_export_vim_h_model.ipynb 教程，就不能自己动手试试吗 把模型按照 结构拆分吗？
export_embedding_model_part_1()
export_embedding_model_part_2()
export_sam_h_model()
onnx_model_example2()
导出onnx 模型后，在导出.engine 模型，python端 在c++端均可。

**Comment by guanhaiye** [comment]
> 请看一下 how_to_export_vim_h_model.ipynb 教程，就不能自己动手试试吗 把模型按照 结构拆分吗？ export_embedding_model_part_1() export_embedding_model_part_2() export_sam_h_model() onnx_model_example2() 导出onnx 模型后，在导出.engine 模型，python端 在c++端均可。

tks
